tests/test_extended_test_case.py

```diff
@@ -137,3 +137,18 @@ def test_assert_selector_with_string(self):
         self.assertContainsSelector(response, "h1", string="Hello World!")
         self.assertNotContainsSelector(response, "h1", string="Hello")
         self.assertContainsSelector(response, "h1", string=re.compile("Hello"))
+
+
+class MessagesTestCase(ExtendedTestCase):
+    def test_assert_has_message(self):
+        url = reverse("message")
+        response = self.client.get(url)
+        message = "I have a message for you."
+        partial_message = "message for you"
+        self.assertIn(partial_message, message)
+        with self.assertRaises(AssertionError):
+            self.assertHasMessage(response, message)
+        response = self.client.get(url, data={"message": message})
+        self.assertHasMessage(response, message)
+        with self.assertRaises(AssertionError):
+            self.assertHasMessage(response, partial_message)
```

Comment:
I try to order tests in given/when/then, separating repeated when/then blocks with newlines:

```suggestion
        message = "I have a message for you."
        partial_message = "message for you"
        url = reverse("message")

        response = self.client.get(url)
        with self.assertRaises(AssertionError):
            self.assertHasMessage(response, message)
        
        response = self.client.get(url, data={"message": message})
        self.assertHasMessage(response, message)
        with self.assertRaises(AssertionError):
            self.assertHasMessage(response, partial_message)
```

Also note I removed `self.assertIn(partial_message, message)`, which just checks the integrity of the test fixtures.